pkg/mutation/schema/schema.go

```diff
@@ -0,0 +1,398 @@
+package schema
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/path/parser"
+	"github.com/open-policy-agent/gatekeeper/pkg/mutation/types"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+// TODO we are assuming that remove will never get corrupted.
+// this should be the case, but is that wise?
+
+// Binding represent the specific GVKs that a
+// mutation's implicit schema applies to
+type Binding struct {
+	Groups   []string
+	Kinds    []string
+	Versions []string
+}
+
+// MutatorWithSchema is a mutator exposing the implied
+// schema of the target object.
+type MutatorWithSchema interface {
+	types.Mutator
+	SchemaBindings() []Binding
+}
+
+var (
+	log = logf.Log.WithName("mutation_schema")
+)
+
+type bindingIterator struct {
+	sorted bool
+}
+
+func (bi *bindingIterator) over(bindings []Binding) func() *schema.GroupVersionKind {
```

Comment:
I think there's no need to have this function as a method. Even if we want to have the sorted flavour, it can be a parameter of the function.

Reply:
Yeah, I think I'm being fancier than I need to be here. I'm coercing everything to run in sorted order. I don't think the time complexity benefits are worth the code complexity cost.